File: apprefurbished/src/main/java/com/androidcollider/easyfin/transactions/list/TransactionsFragment.java
package com.androidcollider.easyfin.transactions.list;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.androidcollider.easyfin.R;
import com.androidcollider.easyfin.common.app.App;
import com.androidcollider.easyfin.common.events.UpdateFrgAccounts;
import com.androidcollider.easyfin.common.events.UpdateFrgHome;
import com.androidcollider.easyfin.common.events.UpdateFrgTransactions;
import com.androidcollider.easyfin.common.managers.resources.ResourcesManager;
import com.androidcollider.easyfin.common.managers.ui.dialog.DialogManager;
import com.androidcollider.easyfin.common.managers.ui.letter_tile.LetterTileManager;
import com.androidcollider.easyfin.common.models.Transaction;
import com.androidcollider.easyfin.common.models.TransactionCategory;
import com.androidcollider.easyfin.common.ui.MainActivity;
import com.androidcollider.easyfin.common.ui.fragments.common.CommonFragmentWithEvents;
import com.androidcollider.easyfin.main.MainFragment;
import com.androidcollider.easyfin.transactions.add_edit.income_expense.AddTransactionIncomeExpenseFragment;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;

/**
 * @author Ihor Bilous
 */

public class TransactionsFragment extends CommonFragmentWithEvents implements TransactionsMVP.View {

    @BindView(R.id.recyclerTransaction)
    RecyclerView recyclerView;
    @BindView(R.id.tvEmptyTransactions)
    TextView tvEmpty;

    private RecyclerTransactionAdapter recyclerAdapter;

    public static final String TRANSACTION = "transaction", MODE = "mode", TYPE = "type";
    public static final int MODE_ADD = 0, MODE_EDIT = 1, TYPE_EXPENSE = 0, TYPE_INCOME = 1;

    @Inject
    ResourcesManager resourcesManager;

    @Inject
    DialogManager dialogManager;

    @Inject
    LetterTileManager letterTileManager;

    @Inject
    TransactionsMVP.Presenter presenter;


    @Override
    public int getContentView() {
        return R.layout.frg_transactions;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((App) getActivity().getApplication()).getComponent().inject(this);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupRecyclerView();
        presenter.setView(this);
        presenter.loadData();
    }

    private void setupRecyclerView() {
        recyclerView.setLayoutManager(new LinearLayoutManager(recyclerView.getContext()));
        recyclerAdapter = new RecyclerTransactionAdapter(resourcesManager, letterTileManager);
        recyclerView.setAdapter(recyclerAdapter);
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {

            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                MainFragment parentFragment = (MainFragment) getParentFragment();
                if (parentFragment != null) {
                    if (dy > 0) {
                        parentFragment.hideMenu();
                    } else if (dy < 0) {
                        parentFragment.showMenu();
                    }
                }
                super.onScrolled(recyclerView, dx, dy);
            }
        });
    }

    private void setVisibility() {
        recyclerView.setVisibility(recyclerAdapter.getItemCount() == 0 ? View.GONE : View.VISIBLE);
        tvEmpty.setVisibility(recyclerAdapter.getItemCount() == 0 ? View.VISIBLE : View.GONE);
    }

    public boolean onContextItemSelected(MenuItem item) {
        int id = recyclerAdapter.getCurrentId();
        switch (item.getItemId()) {
            case R.id.ctx_menu_edit_transaction:
                presenter.getTransactionById(id);
                break;
            case R.id.ctx_menu_delete_transaction:
                showDialogDeleteTransaction(id);
                break;
        }
        return super.onContextItemSelected(item);
    }

    private void showDialogDeleteTransaction(final int id) {
        MainActivity activity = (MainActivity) getActivity();
        if (activity != null) {
            dialogManager.showDeleteDialog(
                    activity,
                    getString(R.string.transaction_delete_warning),
                    (dialog, which) -> presenter.deleteTransactionById(id)
            );
        }
    }

    private void pushBroadcast() {
        EventBus.getDefault().post(new UpdateFrgHome());
        EventBus.getDefault().post(new UpdateFrgAccounts());
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(UpdateFrgTransactions event) {
        presenter.loadData();
    }


    @Override
    public void setTransactionAndTransactionCategoriesLists(List<TransactionViewModel> transactionList,
                                                            List<TransactionCategory> transactionCategoryIncomeList,
                                                            List<TransactionCategory> transactionCategoryExpenseList) {
        recyclerAdapter.setTransactionCategories(transactionCategoryIncomeList, transactionCategoryExpenseList);
        recyclerAdapter.setItems(transactionList);
        setVisibility();
    }

    @Override
    public void goToEditTransaction(Transaction transaction) {
        MainActivity activity = (MainActivity) getActivity();
        if (activity != null) {
            AddTransactionIncomeExpenseFragment frgAddTransDef = new AddTransactionIncomeExpenseFragment();
            Bundle arguments = new Bundle();
            arguments.putInt(MODE, MODE_EDIT);
            arguments.putSerializable(TRANSACTION, transaction);
            frgAddTransDef.setArguments(arguments);
            activity.addFragment(frgAddTransDef);
        }
    }

    @Override
    public void deleteTransaction() {
        recyclerAdapter.deleteItem(recyclerAdapter.getPositionById(recyclerAdapter.getCurrentId()));
        setVisibility();
        pushBroadcast();
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/transactions/list/TransactionsFragment.kt
package com.androidcollider.easyfin.transactions.list

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.core.os.bundleOf
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.androidcollider.easyfin.R
import com.androidcollider.easyfin.common.app.App
import com.androidcollider.easyfin.common.events.UpdateFrgAccounts
import com.androidcollider.easyfin.common.events.UpdateFrgHome
import com.androidcollider.easyfin.common.events.UpdateFrgTransactions
import com.androidcollider.easyfin.common.managers.resources.ResourcesManager
import com.androidcollider.easyfin.common.managers.ui.dialog.DialogManager
import com.androidcollider.easyfin.common.managers.ui.letter_tile.LetterTileManager
import com.androidcollider.easyfin.common.models.Transaction
import com.androidcollider.easyfin.common.models.TransactionCategory
import com.androidcollider.easyfin.common.ui.MainActivity
import com.androidcollider.easyfin.common.ui.fragments.common.CommonFragmentWithEvents
import com.androidcollider.easyfin.main.MainFragment
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import javax.inject.Inject

/**
 * @author Ihor Bilous
 */
class TransactionsFragment : CommonFragmentWithEvents(), TransactionsMVP.View {

    private lateinit var recyclerView: RecyclerView
    private lateinit var tvEmpty: TextView

    private lateinit var recyclerAdapter: RecyclerTransactionAdapter

    @Inject
    lateinit var resourcesManager: ResourcesManager

    @Inject
    lateinit var dialogManager: DialogManager

    @Inject
    lateinit var letterTileManager: LetterTileManager

    @Inject
    lateinit var presenter: TransactionsMVP.Presenter

    override val contentView: Int
        get() = R.layout.frg_transactions

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (activity?.application as App).component?.inject(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupUI(view)
        presenter.setView(this)
        presenter.loadData()
    }

    private fun setupUI(view: View) {
        recyclerView = view.findViewById(R.id.recyclerTransaction)
        tvEmpty = view.findViewById(R.id.tvEmptyTransactions)

        setupRecyclerView()
    }

    private fun setupRecyclerView() {
        recyclerView.layoutManager = LinearLayoutManager(recyclerView.context)
        recyclerAdapter = RecyclerTransactionAdapter(resourcesManager, letterTileManager)
        recyclerView.adapter = recyclerAdapter
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                val parentFragment = parentFragment as MainFragment?
                if (parentFragment != null) {
                    if (dy > 0) {
                        parentFragment.hideMenu()
                    } else if (dy < 0) {
                        parentFragment.showMenu()
                    }
                }
                super.onScrolled(recyclerView, dx, dy)
            }
        })
    }

    private fun setVisibility() {
        recyclerView.visibility = if (recyclerAdapter.itemCount == 0) View.GONE else View.VISIBLE
        tvEmpty.visibility = if (recyclerAdapter.itemCount == 0) View.VISIBLE else View.GONE
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        val id = recyclerAdapter.currentId
        val itemId = item.itemId
        if (itemId == R.id.ctx_menu_edit_transaction) {
            presenter.getTransactionById(id)
        } else if (itemId == R.id.ctx_menu_delete_transaction) {
            showDialogDeleteTransaction(id)
        }
        return super.onContextItemSelected(item)
    }

    private fun showDialogDeleteTransaction(id: Int) {
        val activity = activity as MainActivity?
        if (activity != null) {
            dialogManager.showDeleteDialog(
                activity,
                getString(R.string.transaction_delete_warning)
            ) { presenter.deleteTransactionById(id) }
        }
    }

    private fun pushBroadcast() {
        EventBus.getDefault().post(UpdateFrgHome())
        EventBus.getDefault().post(UpdateFrgAccounts())
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: UpdateFrgTransactions?) {
        presenter.loadData()
    }

    override fun setTransactionAndTransactionCategoriesLists(
        transactionList: List<TransactionViewModel>,
        transactionCategoryIncomeList: List<TransactionCategory>,
        transactionCategoryExpenseList: List<TransactionCategory>
    ) {
        recyclerAdapter.setTransactionCategories(
            transactionCategoryIncomeList,
            transactionCategoryExpenseList
        )
        recyclerAdapter.setItems(transactionList)
        setVisibility()
    }

    override fun goToEditTransaction(transaction: Transaction) {
        findNavController().navigate(
            R.id.addTransactionIncomeExpenseFragment,
            bundleOf(
                MODE to MODE_EDIT,
                TRANSACTION to transaction
            )
        )
    }

    override fun deleteTransaction() {
        recyclerAdapter.deleteItem(recyclerAdapter.getPositionById(recyclerAdapter.currentId))
        setVisibility()
        pushBroadcast()
    }

    companion object {
        const val TRANSACTION = "transaction"
        const val MODE = "mode"
        const val TYPE = "type"
        const val MODE_ADD = 0
        const val MODE_EDIT = 1
        const val TYPE_EXPENSE = 0
        const val TYPE_INCOME = 1
    }
}